### ISSUE TITLE
Understand what's going on with this codebase

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#2955]
This transcript succeeds

```
.> pull git@github.com:pchiusano/unisoncode:2955 .paul
.> pull git@github.com:unisonweb/base:v3:.trunk base.trunk
```

That `2955` branch I created by doing a `push` of my root namespace. 

If instead you do the `pull` of base using my actual codebase (which is the result of doing a codebase upgrade), then the `pull` of trunk fails with:

```
.> pull git@github.com:unisonweb/base:v3:.trunk base.trunk
  Importing downloaded files into local codebase...
Encountered Exception: Sync22Error (HashObjectCorrespondence (ObjectId 3846) [HashId 7170,HashId 7171,HashId 7172,HashId 7173,HashId 7174,HashId 7175,HashId 7176,HashId 7177,HashId 7178] [HashId 119579,HashId 92086,HashId 65167,HashId 91078,HashId 185752,HashId 185753,HashId 185754,HashId 185755,HashId 185756] [ObjectId 76737,ObjectId 60827,ObjectId 60733])
```

It's apparently too big to attach to this issue though.

**Comment by mitchellwrosen** [comment]
The source codebase looks a bit busted:

```
> select count(*) from (select object_id, hash_version, count(*) from hash_object group by object_id, hash_version having count(*) > 1);
2296
```

This query just finds the number of objects that don't have a unique (object_id, hash_id). It should be 0.

**Comment by pchiusano** [comment]
```
select count(*) from (select object_id, hash_version, count(*) from hash_object group by object_id, hash_version having count(*) > 1);
```

returns 2749 for me

**Comment by pchiusano** [comment]
Here's the query to detect object ids with multiple hashes for a given hash version:

```
select count(*) from (select object_id, hash_version, count(*) from hash_object group by object_id, hash_version having count(*) > 1);
```

We did the following:

* Ran the above query on `main` of base. It returns 0 as expected, indicating it's not some historical bug.
* Ran the above query on `v3` of of base, the result of applying migration. It returns 2296, indicating that these objects with more than one v1 hash were somehow introduced after the migration.
* Similar results in the distributed library

The current theory which @ChrisPenner is looking into is that the "sync secondary hashes" logic is busted and is mixing up source and destination hash ids. This would make the bug sort of hard to trigger reliably, since it requires maybe multiple rounds of syncing and depends on hash ids. 

We discussed:

* Just don't sync secondary hashes. This was more of a theoretical feature anyway, to support linking to old hash versions on Unison Share or whatever. But they take up space.
* Maybe add a db constraint to prevent multiple hash_ids for a given (object_id, hash_version). Seems like this should be possible without an aux index - it's just a predicate on the rows being inserted.
* To fix up currently mangled repos, and also probably a good idea anyway, maybe need to implement a gc command that can do the equivalent of syncing the root namespace to a fresh codebase and then making that the current codebase.

**Comment by ChrisPenner** [comment]
On a migrated base there are 14 borked terms and 2360 borked namespaces

Well, depends how you count it, those are the number of objects, there are more hashes.

I believe one cause here might be that we’re mapping all “corrupted” namespaces to the empty branch;
when we re-map object IDs for all these corrupted branches we’d end up with a bunch of previously distinct hashes all being mapped to the empty branch.
There are probably several previously unique histories that are now consolidated as a result, which would cause many namespaces to all map to the same namespaces in the result.
I checked the namespace which had the most hashes mapped to it, which in theory would correspond to the empty branch since every possible divergent history would contain the empty branch at some point, and it was indeed the empty branch :raised_hands:
This explains how we can end up with many v1 hashes mapped to the same object without actually creating any new v1 hashes during the migration

Also, good news, given this situation if we just drop all the old namespace hashes everything will be fine :)

This doesn’t explain the terms though, so it's possible there's something else happening there, but it's possible it's caused by a similar phenomenon where previously unique hashes are consolidated for some reason.

**Comment by ChrisPenner** [comment]
Another update after further digging:

* Looks like the terms which have different hashes in v1 that are consolidated in v2 just had the difference of a top-level Annotation, the new hashing algorithm always adds this if its missing, so it makes sense that they would now have the same hash.
* The corrupted-branch handling is even more complex than originally thought, the consolidation of branches isn't the issue, the issue occurs when a migrated corrupted codebase syncs with a migrated non-corrupted codebase which has the correct values for the hashes which are missing in the other codebase. In this instance, one codebase has mapped v1 hash_objects to the empty branch object, and the other has mapped them to the correct branch value, thus resulting in one hash mapping to two distinct objects.

Given both of these issues, the easiest and probably most correct fix is still to just drop the v1 hash objects entirely.

For codebases which are already out there, Mitchell suggested we bump the schema version and make a new 'migration' that just deletes all v1 hash objects, this seems like a good way to go.

**Comment by ChrisPenner** [comment]
Closed in #2967
